modules/otr.cpp

```diff
@@ -0,0 +1,1291 @@
+            WriteFingerprints();
+        }
+    }
+
+    void CmdDistrust(const CString& sLine) {
+        ConnContext* ctx;
+        Fingerprint* fprint;
+        if (!GetFprintFromArg(sLine, ctx, fprint)) {
+            return;
+        }
+
+        int trusted = otrl_context_is_fingerprint_trusted(fprint);
+        if (!trusted) {
+            PutModuleContext(ctx, CString("Already not trusting ") +
+                                      HumanFingerprint(fprint) + ".");
+        } else {
+            otrl_context_set_trust(fprint, "");
+            PutModuleContext(ctx, CString("Fingerprint ") +
+                                      HumanFingerprint(fprint) +
+                                      " distrusted!");
+            WriteFingerprints();
+        }
+    }
+
+    void CmdFinish(const CString& sLine) {
+        ConnContext* ctx = GetContextFromArg(sLine);
+        if (!ctx) {
+            return;
+        }
+
+        otrl_message_disconnect(m_pUserState, &m_xOtrOps, this,
+                                ctx->accountname, PROTOCOL_ID, ctx->username,
+                                ctx->their_instance);
+        PutModuleContext(ctx, "Conversation finished.");
+    }
+
+    void DoSMP(ConnContext* ctx, const CString& sQuestion,
+               const CString& sSecret) {
+        if (sSecret.Equals("")) {
+            PutModuleContext(ctx, "No secret given!");
+            return;
+        }
+
+        if (ctx->msgstate != OTRL_MSGSTATE_ENCRYPTED) {
+            PutModuleContext(ctx, "Not in OTR session.");
+            return;
+        }
+
+        COtrAppData* ad = static_cast<COtrAppData*>(ctx->app_data);
+        assert(ad);
+
+        if (ad->bSmpReply) {
+            otrl_message_respond_smp(
+                m_pUserState, &m_xOtrOps, this, ctx,
+                reinterpret_cast<const unsigned char*>(sSecret.c_str()),
+                sSecret.length());
+            PutModuleContext(ctx, "Responded to authentication.");
+        } else {
+            if (sQuestion.Equals("")) {
+                otrl_message_initiate_smp(
+                    m_pUserState, &m_xOtrOps, this, ctx,
+                    reinterpret_cast<const unsigned char*>(sSecret.c_str()),
+                    sSecret.length());
+            } else {
+                otrl_message_initiate_smp_q(
+                    m_pUserState, &m_xOtrOps, this, ctx, sQuestion.c_str(),
+                    reinterpret_cast<const unsigned char*>(sSecret.c_str()),
+                    sSecret.length());
+            }
+            PutModuleContext(ctx, "Initiated authentication.");
+        }
+
+        ad->bSmpReply = false;
+    }
+
+    void CmdAuth(const CString& sLine) {
+        ConnContext* ctx = GetContextFromArg(sLine);
+        if (!ctx) {
+            return;
+        }
+
+        CString sSecret = sLine.Token(2, true);
+        DoSMP(ctx, "", sSecret);
+    }
+
+    void CmdAuthQ(const CString& sLine) {
```

Comment:
Users never look at integration test for documentation.
Instructions to run Auth command are sent to user by this module itself, in certain cases.